mongomock/filtering.py

```diff
@@ -408,12 +413,27 @@ def func(doc_val, search_val):
     return func
 
 
-def _type_op(doc_val, search_val):
+def _type_op(doc_val, search_val, in_array=False):
     if search_val not in TYPE_MAP:
         raise OperationFailure('%r is not a valid $type' % search_val)
     elif TYPE_MAP[search_val] is None:
         raise NotImplementedError('%s is a valid $type but not implemented' % search_val)
-    return isinstance(doc_val, TYPE_MAP[search_val])
+
+    spec = TYPE_MAP[search_val]
+    if not isinstance(spec, tuple):
```

Comment:
See my comments below to avoid this step

Reply:
This is still used for lambdas. Unless we want use `'type_name': (lambda ...: ...,)` (which I personally do not find aesthetically appealing 🙂), in that case we could remove this.

Reply:
OK. What do you think of having `TYPE_MAP` being a map with values being predicates instead of being either types or predicates:
```py
def _create_type_predicate(types):
  def _is_a(value):
     return isinstance(value, types)
```

Then:

```py
TYPE_MAP = {
  'double': _create_type_predicate(float),
  'string': _create_type_predicate(str),
  ...
  'null': None,
  ...
  'int': lambda v: ...,
}
```

Reply:
That looks good. In that case, we could use lambdas directly instead of having a new function to create type predicates (e.g. `lambda v: isinstance(v, desired_type_or_types)` instead of  `_create_type_predicate(desired_type_or_types)`). The only problem I see with the sugested approach is with the next patch (in #757), which includes support for refering to other mongo types. For example, the "number" type is defined as in the diff below:

```diff
         isinstance(v, int) and not isinstance(v, bool) and v.bit_length() > 32
     ),
     'decimal': Decimal128,
+    'number': (
+        ('int', 'long', 'double')
+        + (('decimal',) if Decimal128 is not None else ())
+    ),
     'minKey': None,
     'maxKey': None,
 }
```
If we would only work with predicates, that diff above would probably have to be changed to the one below:

```diff
         isinstance(v, int) and not isinstance(v, bool) and v.bit_length() > 32
     ),
     'decimal': Decimal128,
+    'number': lambda v: any((
+        _type_op(v, 'int'),
+        _type_op(v, 'long'),
+        _type_op(v, 'double'),
+        _type_op(v, 'decimal') if Decimal else False,
+    ),
     'minKey': None,
     'maxKey': None,
 }
```

If that's okay for you, let's go for it!

Reply:
Done. Just updated #755.